packages/teams-js/src/public/app.ts

```diff
@@ -806,6 +816,100 @@ export namespace app {
       resolve(sendAndHandleStatusAndReason('executeDeepLink', deepLink));
     });
   }
+
+  /**
+   * Namespace to opt into full-caching or delayed-termination of an app.
+   *
+   * @beta
+   */
+  export namespace lifecycle {
+    /**
+     * Register on resume handler function type
+     *
+     * @param context - Data structure to be used to pass the context to the app.
+     */
+    type registerOnResumeHandlerFunctionType = (context: ResumeContext) => void;
+    /**
+     * A function that is called when the app is ready to be suspended or terminated.
+     */
+    type ReadyToSuspendOrTerminateFunction = () => void;
+
+    /**
+     * Register before suspendOrTerminate handler function type
+     *
+     * @param readyToSuspendOrTerminate - A function that the handler is called with. Apps should call this function when they are ready.
+     * @returns void
+     */
+    type registerBeforeSuspendOrTerminateHandlerFunctionType = (
+      readyToSuspendOrTerminate: ReadyToSuspendOrTerminateFunction,
+    ) => void;
+
+    /**
+     * Registers a handler to be called before the page is suspended or terminatd.
+     *
+     * @param handler - The handler to invoke before the page is suspended or terminated. The page should
+     * invoke the readyToSuspendOrTerminate function provided to it once it's ready.
+     *
+     */
+    export function registerBeforeSuspendOrTerminateHandler(
+      handler: registerBeforeSuspendOrTerminateHandlerFunctionType,
+    ): void {
+      handler && ensureInitialized(runtime);
+      if (handler && !isSupported()) {
```

Comment:
>handler [](http://example.com/codeflow?start=10&length=7)

Similarly to the earlier comment, why do we have to check for `handler` to be defined and non-null here?

Reply:
We don't. Thank you for pointing. I changed the code to explicitly throw error when handler is not provided.

Reply:
Maybe I'm misunderstanding what you are saying -- I don't see any code that throws when`handler` is null/undefined?

Reply:
I am sorry, I had to change it back because unit tests fail if we don't check for handler before calling isSupported. I think the reason is we are assigning null to the handlers in uninitialize call. I have opened another PR to fix it: https://github.com/OfficeDev/microsoft-teams-library-js/pull/1915

Reply:
This is fixed now.